src/cwe_checker_lib/src/analysis/expression_propagation/mod.rs

```diff
@@ -0,0 +1,232 @@
+        &self,
+        value: &Self::Value,
+        _condition: &Expression,
+        _block_before_condition: &Term<Blk>,
+        _is_true: bool,
+    ) -> Option<Self::Value> {
+        Some(value.clone())
+    }
+}
+
+/// Performs the fixpoint algorithm and returns the computation.
+///
+/// Panics, if the computation does not stabilizes.
+fn compute_expression_propagation<'a>(
+    project: &'a Project,
+    graph: &'a Graph,
+) -> Computation<GeneralizedContext<'a, Context<'a>>> {
+    let context = Context::new(project, graph);
+    let mut computation = create_computation(context, None);
+
+    for node in graph.node_indices() {
+        if let Node::BlkStart(_blk, _sub) = graph[node] {
+            // A start in the CFG has no incoming edges in the CFG and
+            // are mostly due to cases where the control flow graph is incomplete.
+            // We assume that no expressions are insertable at such starting nodes.
+            if graph
+                .neighbors_directed(node, petgraph::Incoming)
+                .next()
+                .is_none()
+            {
+                computation.set_node_value(node, NodeValue::Value(HashMap::new()));
+            }
+        }
+    }
+
+    computation.compute_with_max_steps(100);
+
+    if !computation.has_stabilized() {
+        panic!("Fixpoint for expression propagation did not stabilize.");
+    }
+    computation
+}
+
+/// Returns the computed result for every basic block.
+fn extract_results<'a>(
+    graph: &Graph,
+    computation: Computation<GeneralizedContext<'a, Context<'a>>>,
+) -> HashMap<Tid, HashMap<Variable, Expression>> {
+    let mut results = HashMap::new();
+    for node in graph.node_indices() {
+        if let Node::BlkEnd(blk, _sub) = graph[node] {
+            if let Some(NodeValue::Value(insertables)) = computation.get_node_value(node) {
+                results.insert(blk.tid.clone(), insertables.clone());
+            }
+        }
+    }
+    results
+}
+
+/// Replaces for every basic block all propagated expressions.
+///
+/// This uses the expression propagation of basic blocks, thus performs intra-basic-block insertion of expressions.
+fn insert_expressions(
+    inseratables: HashMap<Tid, HashMap<Variable, Expression>>,
+    program: &mut Program,
+) {
+    for sub in program.subs.values_mut() {
+        let mut blocks = sub.term.blocks.iter_mut();
+        // First blocks of functions should not insert any expressions, extracted by the fixpoint computation.
+        if let Some(first_block) = blocks.next() {
+            first_block.merge_def_assignments_to_same_var();
+            first_block.propagate_input_expressions(None);
+        }
+        for block in blocks {
+            block.merge_def_assignments_to_same_var();
+            if let Some(insertable_for_block) = inseratables.get(&block.tid) {
+                block.propagate_input_expressions(Some(insertable_for_block.clone()));
+            }
+        }
+    }
```

Comment:
I am wary of potential undesired interactions of the `merge_def_assignments_to_same_var()` function and the computation of the input expression propagation. The most prudent way would be to use `merge_def_assignments_to_same_var()` before even computing the fixpoint for the expression propagation. This way we could avoid any unexpected interactions that might occur if we do it in another order.